ngraph/core/include/openvino/core/any.hpp

```diff
@@ -0,0 +1,576 @@
+        return *static_cast<decay_t<T>*>(_impl->addressof());
+    }
+
+    /**
+     * Dynamic cast to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <class T>
+    const typename std::enable_if<!std::is_convertible<T, std::shared_ptr<Variant>>::value, T>::type& as() const& {
+        impl_check();
+        _impl->type_check(typeid(decay_t<T>));
+        return *static_cast<const decay_t<T>*>(_impl->addressof());
+    }
+
+    /**
+     * @brief Converts to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    operator T&() & {
+        return as<T>();
+    }
+
+    /**
+     * @brief Converts to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    operator const T&() const& {
+        return as<T>();
+    }
+
+    /**
+     * @brief Converts to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    operator T&() const& {
+        return const_cast<Any*>(this)->as<T>();
+    }
+
+    /**
+     * @brief Converts to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    operator T &&() && {
+        return std::move(as<T&&>());
+    }
+
+    /**
+     * @brief The comparison operator for the Any
+     *
+     * @param other object to compare
+     * @return true if objects are equal
+     */
+    bool operator==(const Any& other) const;
+
+    /**
+     * @brief The comparison operator for the Any
+     *
+     * @param other object to compare
+     * @return true if objects are equal
+     */
+    bool operator==(const std::nullptr_t&) const;
+
+    /**
+     * @brief The comparison operator for the Any
+     *
+     * @param other object to compare
+     * @return true if objects aren't equal
+     */
+    bool operator!=(const Any& other) const;
+
+    /**
+     * @brief Standard pointer operator
+     * @return underlined interface
+     */
+    Base* operator->();
+
+    /**
+     * @brief Standard pointer operator
+     * @return underlined interface
+     */
+    const Base* operator->() const;
+
+    /**
+     * @brief Prints underlying object to the given output stream.
+     * Uses operator<< if it is defined, leaves stream unchanged otherwise.
+     * In case of empty any or nullptr stream immediately returns.
+     *
+     * @param stream Output stream object will be printed to.
+     */
+    void print(std::ostream& stream) const;
+
+    /**
+     * @brief Return pointer to underlined interface
+     * @return underlined interface
+     */
+    Base* get() {
+        impl_check();
+        return _impl.get();
+    }
+
+    /**
+     * @brief Return pointer to underlined interface
+     * @return underlined interface
+     */
+    const Base* get() const {
+        impl_check();
+        return _impl.get();
+    }
+};
+
+namespace util {
+template <>
+struct AsTypePtr<Any> {
```

Comment:
Why do we need this structure?

Reply:
To customize `ov::as_type_ptr` behavior for ov::Any. Template Type specialization is march more simple than template function specialization and much more customizable.